editor/src/components/navigator/navigator-item/navigator-item-wrapper.tsx

```diff
@@ -25,16 +30,92 @@ interface NavigatorItemWrapperProps {
+  (store: EditorStore) => store.derived.canvas.transientState,
+  (store) => store.derived.navigatorTargets,
+  (store) => store.derived.elementWarnings,
+  (store, templatePath) =>
+    TP.isScenePath(templatePath)
+      ? null
+      : MetadataUtils.getElementByInstancePathMaybe(store.editor.jsxMetadataKILLME, templatePath),
+  (store) =>
+    store.derived.canvas.transientState.fileState == null
+      ? getOpenImportsFromState(store.editor)
+      : store.derived.canvas.transientState.fileState.imports,
+  (_: EditorStore, templatePath: TemplatePath) => templatePath,
+  (
+    jsxMetadataKILLME,
+    transientState,
+    navigatorTargets,
+    elementWarnings,
+    element,
+    imports,
+    templatePath: TemplatePath,
+  ) => {
+    const fileState = transientState.fileState
+
+    const componentsIncludingScenes: Array<UtopiaJSXComponent> =
+      fileState == null
+        ? []
+        : fileState.topLevelElementsIncludingScenes.filter(isUtopiaJSXComponent)
+    const elementOriginType = MetadataUtils.getElementOriginType(
+      componentsIncludingScenes,
+      jsxMetadataKILLME,
+      templatePath,
+    )
+    const staticName = MetadataUtils.getStaticElementName(
+      templatePath,
+      componentsIncludingScenes,
+      jsxMetadataKILLME,
+    )
+    const labelInner = MetadataUtils.getElementLabel(templatePath, jsxMetadataKILLME, staticName)
+    const componentInstanceInner = MetadataUtils.isComponentInstance(
+      templatePath,
+      componentsIncludingScenes,
+      jsxMetadataKILLME,
+      imports,
+    )
+    // FIXME: This is a mitigation for a situation where somehow this component re-renders
+    // when the navigatorTargets indicate it shouldn't exist...
+    const isInNavigatorTargets = TP.containsPath(templatePath, navigatorTargets)
+    let noOfChildrenInner: number = 0
+    let supportsChildren: boolean = false
+    if (isInNavigatorTargets) {
+      noOfChildrenInner = MetadataUtils.getImmediateChildren(jsxMetadataKILLME, templatePath).length
+      supportsChildren = MetadataUtils.targetSupportsChildren(
+        imports,
+        jsxMetadataKILLME,
+        templatePath,
+      )
+    }
+
+    const elementWarningsInner = getValueFromComplexMap(TP.toString, elementWarnings, templatePath)
+
+    return {
+      staticElementName: staticName,
+      label: labelInner,
+      element: element,
+      componentInstance: componentInstanceInner,
+      isAutosizingView: MetadataUtils.isAutoSizingView(element),
+      isSelected: TP.containsPath(templatePath, transientState.selectedViews),
+      isHighlighted: TP.containsPath(templatePath, transientState.highlightedViews),
+      noOfChildren: noOfChildrenInner,
+      supportsChildren: supportsChildren,
+      imports: imports,
+      elementOriginType: elementOriginType,
+      elementWarnings: elementWarningsInner ?? defaultElementWarnings,
+    }
+  },
+)((store, templatePath: TemplatePath) => TP.toString(templatePath))
```

Comment:
this is the re-reselect part. you need to pass in a function that generates a cache key. here I use the one property (templatePath) that distinguishes the NavigatorWrapperItems from each other.